lib/spack/spack/variant.py

```diff
@@ -254,116 +239,143 @@ def _flatten(values) -> Collection:
 
 
 #: Type for value of a variant
-ValueType = Union[str, bool, Tuple[Union[str, bool], ...]]
+ValueType = Tuple[Union[bool, str], ...]
 
 #: Type of variant value when output for JSON, YAML, etc.
 SerializedValueType = Union[str, bool, List[Union[str, bool]]]
 
 
 @lang.lazy_lexicographic_ordering
-class VariantBase:
-    """A BaseVariant corresponds to a spec string of the form ``foo=bar`` or ``foo=bar,baz``.
-    It is a constraint on the spec and abstract in the sense that it must have **at least** these
-    values -- concretization may add more values."""
+class VariantValue:
+    """A VariantValue is a key-value pair that represents a variant. It can have zero or more
+    values. Values have set semantics, so they are unordered and unique. The variant type can
+    be narrowed from multi to single to boolean, this limits the number of values that can be
+    stored in the variant. Multi-valued variants can either be concrete or abstract: abstract
+    means that the variant takes at least the values specified, but may take more when concretized.
+    Concrete means that the variant takes exactly the values specified. Lastly, a variant can be
+    marked as propagating, which means that it should be propagated to dependencies."""
 
     name: str
     propagate: bool
-    _value: ValueType
-    _original_value: Any
+    concrete: bool
+    type: VariantType
+    _values: ValueType
 
-    def __init__(self, name: str, value: ValueType, propagate: bool = False) -> None:
+    def __init__(
+        self,
+        name: str,
+        value: ValueType,
+        type: VariantType,
```

Comment:
(nitpick) Can we use another name, e.g. `variant_type`, to avoid shadowing a built-in function?